solidity/random-beacon/contracts/DKGValidator.sol

```diff
@@ -0,0 +1,230 @@
+    ///         and signatures of operators supporting the result.
+    /// @param seed seed used to start the DKG and select group members
+    /// @param startBlock DKG start block
+    /// @return isValid true if the result is valid, false otherwise
+    /// @return errorMsg validation error message; empty for a valid result
+    function validate(
+        DKG.Result calldata result,
+        uint256 seed,
+        uint256 startBlock
+    ) external view returns (bool isValid, string memory errorMsg) {
+        (bool hasValidFields, string memory error) = validateFields(result);
+        if (!hasValidFields) {
+            return (false, error);
+        }
+
+        if (!validateSignatures(result, startBlock)) {
+            return (false, "Invalid signatures");
+        }
+
+        if (!validateGroupMembers(result, seed)) {
+            return (false, "Invalid group members");
+        }
+
+        return (true, "");
+    }
+
+    /// @notice Performs a static validation of DKG result fields: lengths,
+    ///         ranges, and order of arrays.
+    /// @return isValid true if the result is valid, false otherwise
+    /// @return errorMsg validation error message; empty for a valid result
+    function validateFields(DKG.Result calldata result)
+        public
+        view
+        returns (bool isValid, string memory errorMsg)
+    {
+        // Group public key needs to be 128 bytes long.
+        if (result.groupPubKey.length != 128) {
+            return (false, "Malformed group public key");
+        }
+
+        // The number of misbehaved members can not exceed the threshold.
+        // Misbehaved member indices needs to be unique, between [1,64],
+        // and sorted in ascending order.
+        uint8[] calldata misbehavedMembersIndices = result
+            .misbehavedMembersIndices;
+        if (groupSize - misbehavedMembersIndices.length < activeThreshold) {
+            return (false, "Too many members misbehaving during DKG");
+        }
+        if (misbehavedMembersIndices.length > 1) {
+            if (
+                misbehavedMembersIndices[0] < 1 ||
+                misbehavedMembersIndices[misbehavedMembersIndices.length - 1] >
+                64
+            ) {
+                return (false, "Corrupted misbehaved members indices");
+            }
+            for (uint256 i = 1; i < misbehavedMembersIndices.length; i++) {
+                if (
+                    misbehavedMembersIndices[i - 1] >=
+                    misbehavedMembersIndices[i]
+                ) {
+                    return (false, "Corrupted misbehaved members indices");
+                }
+            }
+            uint8 last = misbehavedMembersIndices[
+                misbehavedMembersIndices.length - 1
+            ];
+            if (last < 1 || last > groupSize) {
+                return (false, "Corrupted misbehaved members indices");
+            }
```

Comment:
We do the same for `signingMembersIndices`. Maybe it would be worth extracting this block to separate function and improve readability? Something like `validateMembersIndices` in #2736

Reply:
`misbehavedMembersIndices` is `uint8[]` and `signingMembersIndices` is `uint256[]` (for gas efficiency reasons) :(

Reply:
Ah ok. But, can't we convert `signingMembersIndices` to `uint8[]` as well? Max index is `64` and we could save on gas in the same way?

Reply:
I am pretty sure it was something about gas cost. If we make then `uint8[]`, the compiler converts to `uint256` when we use them. Let's compare the gas cost once we merge this PR and change (or not).